### PR TITLE
Refactored igemm kernel and solver for xdlops wrw fp32

### PR DESCRIPTION
**Pull request by zjing14**
- Dedicated wrw solver with reduced tuning parameters
- Dedicated wrw kernel instead of reusing fwd kernel
- Changed loading dimension of matrix B from GemmN (CYX) to GemmK (NHW) for enabling vector load (up to 15% improvement)
- Do not support group conv. in this PR

### CODE REVIEW

**Comment by daniellowell** [comment]
@zjing14 Does this retuning? If so, what needs retuning?

**Comment by zjing14** [comment]
> @zjing14 Does this retuning? If so, what needs retuning?

Yes, we need to retune wrw fp32.

**Comment by atamazov** [comment]
@zjing14 Am I right that this new solver is factored out from some other "universal" solver? If it is expected to be always better with the _subset_ of problem-configs it covers (the non-group WrW convolutions, I guess), then I recommend excluding this _subset_ from that "universal" solver.

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_K_PER_BLOCK=") + std::to_string(config.GemmKPerBlock) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_K_BLOCKS=") + std::to_string(config.GemmKBlocks) +
+        std::string(" -DCK_PARAM_DEPENDENT_GRID_SIZE=") + std::to_string(grid_size) +
+        std::string(" -DCK_PARAM_GEMM_M_PER_WAVE=") + std::to_string(config.GemmMPerWave) +
+        std::string(" -DCK_PARAM_GEMM_N_PER_WAVE=") + std::to_string(config.GemmNPerWave) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_B_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_K=") + std::to_string(GemmBBlockCopyClusterLengths_GemmK) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_B_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_N=") + std::to_string(GemmBBlockCopyClusterLengths_GemmN) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_A_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_K=") + std::to_string(GemmABlockCopyClusterLengths_GemmK) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_A_BLOCK_COPY_CLUSTER_LENGTHS_GEMM_M=") + std::to_string(GemmABlockCopyClusterLengths_GemmM) +
+        std::string(" -DCK_PARAM_TUNABLE_GEMM_B_BLOCK_COPY_SRC_DATA_PER_READ_GEMM=") + std::to_string(GemmBBlockCopySrcDataPerRead_GemmK) +
```
[the remark this answers is not in the document]
Comment (zjing14) [reply src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp:530]: As the loading dimension is not fixed in the future, we intend to remove the suffix.

File: src/solver/implicitgemm_util.hpp [hunk 2]
```diff
@@ -423,6 +423,48 @@ inline static uint32_t GetEPackLength(const ConvolutionContext& ctx, bool isXdlo
...
+        return false;
+    if(GemmMPerWave == 32 && GemmNPerWave == 32 && GemmKPackedPerBlock % 2 != 0)
+        return false;
+    if(GemmMPerWave == 16 && GemmNPerWave == 16 && GemmKPackedPerBlock % 4 != 0)
+        return false;
+
+    const auto WaveSize  = 64;
+    const auto BlockSize = GemmNPerBlock * GemmMPerBlock / (GemmMPerWave * GemmNPerWave) * WaveSize;
+
+    if(BlockSize < 64 || BlockSize > 512)
```
[the remark this answers is not in the document]
Comment (zjing14) [reply src/solver/implicitgemm_util.hpp]: I tested the forward direction, enlarging blockSize to 512 can help performance. But we need more tests and leave the change to another PR.

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    const std::size_t n  = ConvolutionContextInterpreter::GetBatchN(ctx);
+    const std::size_t k  = ConvolutionContextInterpreter::GetOutputChannelK(ctx) / ctx.group_counts;
+    const std::size_t c  = ConvolutionContextInterpreter::GetInputChannelC(ctx) / ctx.group_counts;
+    const std::size_t y  = ConvolutionContextInterpreter::GetFilterHeightY(ctx);
+    const std::size_t x  = ConvolutionContextInterpreter::GetFilterWidthX(ctx);
+    const std::size_t ho = ConvolutionContextInterpreter::GetOutputHeightHo(ctx);
+    const std::size_t wo = ConvolutionContextInterpreter::GetOutputWidthWo(ctx);
+
+    const std::size_t GemmM = k;
+    const std::size_t GemmN = n * ho * wo;
```
Comment (TejashShah) [inline src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp]: GemmM, GemmN and GemmK formulae are wrong

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TejashShah) [reply src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp:530]: are you saying that we are going to make it tunable? If yes, I agree with your reasoning. Otherwise, it is more readable to put the dimension name like GemmK as suffix

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp (hunk 3 above)
Comment (zjing14) [reply src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp]: Fixed. Thanks for finding it out.

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zjing14) [reply src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp:530]: Yes. We need to make it tunable in the future.

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+             GemmABlockCopyDstDataPerWrite_GemmM,
+             std::ignore) = config.CalculateGemmABlockCopyPerformanceParameters(ctx);
+
+    std::tie(GemmBBlockCopyClusterLengths_GemmK,
+             GemmBBlockCopyClusterLengths_GemmN,
+             GemmBBlockCopySrcDataPerRead_GemmK,
+             GemmBBlockCopyDstDataPerWrite_GemmN,
+             std::ignore) = config.CalculateGemmBBlockCopyPerformanceParameters(ctx);
+
+    construction_parameters.kernel_file = "gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_"
```
Comment (TejashShah) [inline src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp:489]: can you please not split this kernel name/file name into 2 separate lines? Searching for this string in grep makes it difficult. I believe, clang-format will be splitting into 2 for you. You could disable clang-format by 

// clang-format off

// clang-format on

File: src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
+#include "common_header.hpp"
```
Comment (TejashShah) [inline src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp:1]: By creating this new file, we have caused more code duplication. There was need to create a separate wrw solver as it has reduced search space and slightly different tuning parameters. But this file contents' could have remained in the old file

File: src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+#include "float_types.h"
+
+extern "C" __global__
+    __launch_bounds__(CK_PARAM_TUNABLE_BLOCK_SIZE, 2) void gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer(
+        const FLOAT* const __restrict__ p_in_global,
+        const FLOAT* const __restrict__ p_out_global,
+        FLOAT* const __restrict__ p_wei_global)
+{
+#if !(MIOPEN_USE_FP32 && CK_PARAM_PROBLEM_CONV_DIRECTION_FORWARD)
+    static_assert(false, "Only support forward fp32!");
```
Comment (TejashShah) [inline src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp:13]: Please correct the error message.. wrw fp32

File: src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp [hunk 7]
```diff
@@ -10,7 +10,7 @@ extern "C" __global__
         FLOAT* const __restrict__ p_wei_global)
 {
 #if !(MIOPEN_USE_FP32 && CK_PARAM_PROBLEM_CONV_DIRECTION_FORWARD)
```
Comment (TejashShah) [inline src/kernels/composable_kernel/src/kernel_wrapper/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp:12]: CK_PARAM_PROBLEM_CONV_DIRECTION_FORWARD

change to 

CK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT

Alongside, we need to change solver to set 
CK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT=1
It is currently set CK_PARAM_PROBLEM_CONV_DIRECTION_FORWARD=0

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+
+    // clang-format off
+    construction_parameters.kernel_file = "gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer.cpp";
+    construction_parameters.kernel_name = "gridwise_convolution_implicit_gemm_v4r4_gen_xdlops_wrw_fp32_nchw_kcyx_nkhw_lds_double_buffer";
+
+    construction_parameters.comp_options =
+        std::string(" -std=c++14 ") +
+        std::string(" -DCK_PARAM_PROBLEM_CONV_DIRECTION_FORWARD=") + std::to_string(1) +
+        std::string(" -DCK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_DATA=") + std::to_string(0) +
+        std::string(" -DCK_PARAM_PROBLEM_CONV_DIRECTION_BACKWARD_WEIGHT=") + std::to_string(0) +
```
Comment (TejashShah) [inline src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops_wrw_fp32.cpp:497]: BACKWARD_WEIGHT=1
FORWARD=0

https://github.com/ROCmSoftwarePlatform/MIOpen/pull/152/files#r414217587

**Comment by TejashShah** [comment]
@zjing14 Sorry, I am raising comment after comment as I discover while making fp16/bfp16 specific wrw solver.

**Comment by daniellowell** [comment]
@zjing14 Merge conflicts

**Comment by zjing14** [comment]
@daniellowell Merged.

File: src/include/miopen/solver.hpp [hunk 9]
```diff
@@ -935,6 +935,75 @@ struct ConvHipImplicitGemmV4R4GenFwdXdlops : SolverBase<ConvolutionContext>
...
+    GetPerformanceConfig(const ConvolutionContext& ctx) const;
+    bool IsValidPerformanceConfig(const ConvolutionContext& ctx,
+                                  const PerformanceImplicitGemmV4R4GenXdlopsWrWFp32& c) const;
+    bool IsApplicable(const ConvolutionContext& ctx) const;
+    ConvSolution GetSolution(const ConvolutionContext& ctx,
+                             const PerformanceImplicitGemmV4R4GenXdlopsWrWFp32& config,
+                             bool disableConfigOverrideFromEnv = false) const;
+
+    PerformanceImplicitGemmV4R4GenXdlopsWrWFp32 Search(const ConvolutionContext&) const;
+    int RunAndMeasureSolution(miopen::Handle& profile_h,
```
Comment (atamazov) [inline src/include/miopen/solver.hpp:997]: [anchor]

**Comment by atamazov** [comment]
This is a blocker for #203 due to [this](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/152#discussion_r422289511). It MUST be reverted and reworked to invokers.

Since #145 we should NOT use `RunAndMeasure()` in the new Solvers. The only reason that this PR has passed the tests is that #203 is not yet merged.

@daniellowell @DrizztDoUrden @alexandraBara @JehandadKhan @zjing14

**Comment by atamazov** [comment]
The most recent comment is wrong.

File: src/solver.cpp [hunk 10]
```diff
@@ -307,6 +307,11 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
                        ++id,
                        ConvHipImplicitGemmV4R4GenXdlopsFwdFp32{},
                        miopenConvolutionAlgoImplicitGEMM);
+
+    RegisterWithSolver(registry,
```
Comment (atamazov) [inline src/solver.cpp:311]: I think we shall revert this change...

File: src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp [hunk 11]
```diff
@@ -402,6 +402,10 @@ bool ConvHipImplicitGemmV4R4GenWrWXdlops::IsApplicable(const ConvolutionContext&
     if(!(ctx.IsFp32() || ctx.IsFp16() || ctx.IsBfp16()))
         return false;
 
+    // covered by ConvHipImplicitGemmV4R4GenXdlopsWrWFp32
```
Comment (atamazov) [inline src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp:405]: ...and this change.
